SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -526,22 +535,48 @@ void USpatialSender::SendRPC(TSharedRef<FPendingRPCParams> Params)
 	}
 }
 
-void USpatialSender::EnqueueRetryRPC(TSharedRef<FPendingRPCParams> Params)
+void USpatialSender::EnqueueRetryRPC(TSharedRef<FReliableRPCForRetry> RetryRPC)
 {
-	RetryRPCs.Add(Params);
+	RetryRPCs.Add(RetryRPC);
 }
 
 void USpatialSender::FlushRetryRPCs()
 {
 	// Retried RPCs are sorted by their index.
-	RetryRPCs.Sort([](const TSharedPtr<FPendingRPCParams>& A, const TSharedPtr<FPendingRPCParams>& B) { return A->RetryIndex < B->RetryIndex; });
+	RetryRPCs.Sort([](const TSharedRef<FReliableRPCForRetry>& A, const TSharedRef<FReliableRPCForRetry>& B) { return A->RetryIndex < B->RetryIndex; });
 	for (auto& RetryRPC : RetryRPCs)
 	{
-		SendRPC(RetryRPC);
+		RetryReliableRPC(RetryRPC);
 	}
 	RetryRPCs.Empty();
 }
 
+void USpatialSender::RetryReliableRPC(TSharedRef<FReliableRPCForRetry> RetryRPC)
+{
+	if (!RetryRPC->TargetObject.IsValid())
+	{
+		// Target object was destroyed before the RPC could be (re)sent
+		return;
+	}
+
+	UObject* TargetObject = RetryRPC->TargetObject.Get();
+	FUnrealObjectRef TargetObjectRef = PackageMap->GetUnrealObjectRefFromObject(TargetObject);
+	if (TargetObjectRef == FUnrealObjectRef::UNRESOLVED_OBJECT_REF)
+	{
+		UE_LOG(LogSpatialSender, Warning, TEXT("Actor %s got unresolved (?) before RPC %s could be retried. This RPC will not be sent."), *TargetObject->GetName(), *RetryRPC->Function->GetName());
+		return;
+	}
+
+	Worker_CommandRequest CommandRequest = CreateRetryRPCCommandRequest(*RetryRPC, TargetObjectRef.Offset);
```

Comment:
Any reason the rpc's object offset couldn't be stored in the RetryRPC itself?

Reply:
Entity ID and object offset are "contained" in the target object in the sense that we can look up its object ref and get them from there. Alternatively, we could store the entity ID and offset in the struct, and at that point we would only need to store `TargetObject` to check that it's still valid. I considered removing that logic (checking for target object validity), but came to the conclusion it's still needed because clients can try to send RPCs on actors just as server is deleting them, in which case there's no point for clients to keep spamming retries.